File: packages/sdk-ts/src/adapters/redis.ts
// Redis adapter for `VerifyCache`. Duck-typed against any client that
// exposes `get(key) → string|Buffer|null`, `setex(key, seconds, value)`
// or `set(key, value, "EX", seconds)`, and `del(key)`. Works with
// `ioredis`, `node-redis`, and the Upstash REST shim out of the box.
//
// Design (validated against the AEGIS server-side Redis pattern at
// apps/api/src/common/redis/redis.service.ts):
//
//   * Fail-soft. A Redis miss, timeout, or decode error returns
//     `undefined` — the gateway then falls through to the network.
//     We NEVER let a backend wobble cascade into denied verifies.
//   * TTL in seconds (Redis native). Adapter computes
//     `ceil((expiresAt − now) / 1000)` from the gateway's epoch-ms.
//   * Key namespace: `aegis:verify:<sha256>` — matches the existing
//     `namespace:resource:id` colon convention.
//   * `onError` hook surfaces backend errors so operators can alarm
//     without having to touch the gateway's `onError` (cache backend
//     errors are not AegisErrors).
//
// No hard dependency on any Redis client — package consumers install
// their preferred client themselves.

import type { CachedVerify, VerifyCache } from '../cache.js';
import type { VerifyResult } from '../types.js';

/** Minimal Redis surface this adapter relies on. */
export interface RedisLike {
  get(key: string): Promise<string | null | undefined>;
  // Two common signatures across clients:
  //   ioredis:        set(key, value, "EX", seconds)
  //   node-redis v4:  set(key, value, { EX: seconds })
  // We expose both for the adapter's own internal `setEx` helper.
  set?(key: string, value: string, ...args: unknown[]): Promise<unknown>;
  setex?(key: string, seconds: number, value: string): Promise<unknown>;
  del(key: string | string[]): Promise<unknown>;
}

export interface RedisVerifyCacheOptions {
  /** Key prefix. Default: `'aegis:verify:'` (matches server convention). */
  keyPrefix?: string;
  /**
   * Backend error hook. Fires when Redis throws or returns malformed
   * data. Use for metrics/alarms. Must not throw. Backend errors are
   * NEVER propagated to the verify call — they degrade to cache miss.
   */
  onError?: (op: 'get' | 'set' | 'delete', err: unknown, key: string) => void;
  /** Clock injection for tests. Default: `Date.now`. */
  now?: () => number;
}

const DEFAULT_PREFIX = 'aegis:verify:';

export class RedisVerifyCache implements VerifyCache {
  private readonly client: RedisLike;
  private readonly prefix: string;
  private readonly onError: ((op: 'get' | 'set' | 'delete', err: unknown, key: string) => void) | undefined;
  private readonly now: () => number;

  constructor(client: RedisLike, opts: RedisVerifyCacheOptions = {}) {
    this.client = client;
    this.prefix = opts.keyPrefix ?? DEFAULT_PREFIX;
    this.onError = opts.onError;
    this.now = opts.now ?? Date.now;
  }

  async get(key: string): Promise<CachedVerify | undefined> {
    const fullKey = this.prefix + key;
    let raw: string | null | undefined;
    try {
      raw = await this.client.get(fullKey);
    } catch (err) {
      this.safeOnError('get', err, fullKey);
      return undefined;
    }
    if (raw === null || raw === undefined) return undefined;
    return this.decode(raw, fullKey);
  }

  /**
   * Redis enforces TTL natively, so peek and get are equivalent at the
   * backend — Redis already evicted anything past its TTL. The gateway
   * still validates `expiresAt` itself for the half-open serve-stale
   * window, but that window is bounded by the server's `ttl` field.
   */
  async peek(key: string): Promise<CachedVerify | undefined> {
    return this.get(key);
  }

  async set(key: string, value: CachedVerify): Promise<void> {
    const fullKey = this.prefix + key;
    const ttlSec = Math.max(1, Math.ceil((value.expiresAt - this.now()) / 1000));
    let payload: string;
    try {
      payload = JSON.stringify({ result: value.result, expiresAt: value.expiresAt });
    } catch (err) {
      this.safeOnError('set', err, fullKey);
      return;
    }
    try {
      if (typeof this.client.setex === 'function') {
        await this.client.setex(fullKey, ttlSec, payload);
        return;
      }
      if (typeof this.client.set === 'function') {
        // ioredis: set(key, value, "EX", seconds)
        // node-redis v4: set(key, value, { EX: seconds })
        // Try ioredis-style first (more common in AEGIS' own apps/api).
        await this.client.set(fullKey, payload, 'EX', ttlSec);
        return;
      }
      throw new TypeError('RedisLike client must implement set or setex');
    } catch (err) {
      this.safeOnError('set', err, fullKey);
    }
  }

  async delete(key: string): Promise<void> {
    const fullKey = this.prefix + key;
    try {
      await this.client.del(fullKey);
    } catch (err) {
      this.safeOnError('delete', err, fullKey);
    }
  }

  private decode(raw: string, fullKey: string): CachedVerify | undefined {
    try {
      const parsed = JSON.parse(raw) as { result: VerifyResult; expiresAt: number };
      if (
        parsed === null ||
        typeof parsed !== 'object' ||
        typeof parsed.expiresAt !== 'number' ||
        parsed.result === undefined
      ) {
        // type-rationale: defensive — a corrupt payload should degrade
        // to miss, not crash the gateway. Surface via onError so the
        // operator notices.
        throw new TypeError('Malformed VerifyCache payload');
      }
      return { result: parsed.result, expiresAt: parsed.expiresAt };
    } catch (err) {
      this.safeOnError('get', err, fullKey);
      return undefined;
    }
  }

  private safeOnError(op: 'get' | 'set' | 'delete', err: unknown, key: string): void {
    if (!this.onError) return;
    try {
      this.onError(op, err, key);
    } catch {
      // Hooks must never break the verify path.
    }
  }
}

File: packages/sdk-py/aegis/adapters/redis.py
"""Redis adapter for ``VerifyCache``. Mirrors
``packages/sdk-ts/src/adapters/redis.ts``.

Duck-typed against any client that exposes async ``get`` / ``set`` /
``delete`` (matching ``redis.asyncio.Redis`` from redis-py 5.x or
anything that quacks like it). No hard dependency on redis-py — the
package consumer installs their preferred client.

Design contract — validated against
``apps/api/src/common/redis/redis.service.ts``:

* **Fail-soft.** Redis miss / timeout / decode error returns ``None``;
  the gateway then falls through to the network. Backend wobbles never
  cascade into denied verifies.
* **TTL in seconds** (Redis native). Adapter computes
  ``ceil((expires_at_ms − now) / 1000)`` from the gateway's epoch-ms.
* **Key namespace**: ``aegis:verify:<sha256>`` — matches the existing
  ``namespace:resource:id`` colon convention.
* **on_error hook** surfaces backend errors so operators can alarm.
"""

from __future__ import annotations

import json
import math
from time import time_ns
from typing import Any, Awaitable, Callable, Literal, Optional, Protocol, runtime_checkable

from ..models import VerifyResult
from ..verify_cache import CachedVerify

DEFAULT_PREFIX = "aegis:verify:"

CacheOp = Literal["get", "set", "delete"]


@runtime_checkable
class RedisLike(Protocol):
    """Minimal async Redis surface this adapter relies on. Matches
    ``redis.asyncio.Redis`` and most async Upstash shims.
    """

    def get(self, key: str) -> Awaitable[Optional[bytes | str]]: ...
    def set(
        self,
        key: str,
        value: str | bytes,
        *,
        ex: int | None = ...,
    ) -> Awaitable[Any]: ...
    def delete(self, *keys: str) -> Awaitable[Any]: ...


class RedisVerifyCache:
    """Async Redis-backed ``VerifyCache`` for ``AsyncVerifyGateway``."""

    __slots__ = ("_client", "_prefix", "_on_error", "_now")

    def __init__(
        self,
        client: Any,  # duck-typed; constrained by RedisLike Protocol at usage
        *,
        key_prefix: str = DEFAULT_PREFIX,
        on_error: Callable[[CacheOp, BaseException, str], None] | None = None,
        now: Callable[[], int] | None = None,
    ) -> None:
        self._client = client
        self._prefix = key_prefix
        self._on_error = on_error
        self._now = now or (lambda: time_ns() // 1_000_000)

    async def get(self, key: str) -> CachedVerify | None:
        full_key = self._prefix + key
        try:
            raw = await self._client.get(full_key)
        except Exception as err:  # noqa: BLE001 — fail-soft contract
            self._safe_on_error("get", err, full_key)
            return None
        if raw is None:
            return None
        return self._decode(raw, full_key)

    async def peek(self, key: str) -> CachedVerify | None:
        # Redis enforces TTL natively — peek and get are equivalent at
        # the backend. The gateway still validates expires_at itself for
        # the half-open serve-stale window.
        return await self.get(key)

    async def set(self, key: str, value: CachedVerify) -> None:
        full_key = self._prefix + key
        ttl_sec = max(1, math.ceil((value.expires_at_ms - self._now()) / 1000))
        try:
            payload = json.dumps(
                {
                    "result": value.result.model_dump(by_alias=True, mode="json"),
                    "expiresAt": value.expires_at_ms,
                }
            )
        except Exception as err:  # noqa: BLE001
            self._safe_on_error("set", err, full_key)
            return
        try:
            # ``ex=`` is the canonical kwarg in redis-py 5.x. Most async
            # shims (aioredis 2.x, Upstash) accept it too.
            await self._client.set(full_key, payload, ex=ttl_sec)
        except Exception as err:  # noqa: BLE001
            self._safe_on_error("set", err, full_key)

    async def delete(self, key: str) -> None:
        full_key = self._prefix + key
        try:
            await self._client.delete(full_key)
        except Exception as err:  # noqa: BLE001
            self._safe_on_error("delete", err, full_key)

    def _decode(self, raw: bytes | str, full_key: str) -> CachedVerify | None:
        try:
            text = raw.decode("utf-8") if isinstance(raw, (bytes, bytearray)) else raw
            parsed = json.loads(text)
            if (
                not isinstance(parsed, dict)
                or "result" not in parsed
                or not isinstance(parsed.get("expiresAt"), int)
            ):
                raise TypeError("malformed VerifyCache payload")
            result = VerifyResult.model_validate(parsed["result"])
            return CachedVerify(result=result, expires_at_ms=parsed["expiresAt"])
        except Exception as err:  # noqa: BLE001
            self._safe_on_error("get", err, full_key)
            return None

    def _safe_on_error(self, op: CacheOp, err: BaseException, key: str) -> None:
        if self._on_error is None:
            return
        try:
            self._on_error(op, err, key)
        except Exception:  # noqa: BLE001 — hooks must never break the verify path
            pass
